--- a/fs/fuse/fuse_i.h
+++ b/fs/fuse/fuse_i.h
@@ -587,10 +587,13 @@ struct fuse_fs_context {
 	const char *subtype;
 	struct bpf_prog *root_bpf;
 	struct file *root_dir;
 
+	/* ANDROID: fuse-bpf backport - cached resolved file for ctx->fd */
+	struct file *file;
+
 	/* DAX device, may be NULL */
 	struct dax_device *dax_dev;
 
 	/* fuse_dev pointer to fill in, should contain NULL on entry */
 	void **fudptr;
 };
